Add `output` property to MappedOperator

josh-fell opened the pull request:
Currently the `output` property of operators is only available to unmapped "classic" operators. There can be use cases in which `XComs` from a set of mapped tasks need to be consumed by a downstream task. Of course, accessing these mapped-task `XComs` can be done with the typical Jinja template, but having the convenience of using the `output` property would be a "quality of life" improvement for DAG authors (especially since task dependencies would then be automatically implemented too).

~Rather than adding the property to `MappedOperator` and have this property definition exist in two locations (along with `BaseOperator`), I opted to add this to `AbstractOperator` given all tasks have an output even if that output is empty and both `BaseOperator` and `MappedOperator` share the `AbstractOperator` implementation. If there a higher-level reason to have two instances of the property definition, I'm happy to oblige.~

This PR adds the `output` property to `MappedOperator` and includes a return type annotation for `XComArg`. Additionally, until a more appropriate API implementation of accessing a different `XCom` key from `output` other than "return_value" is decided, the example DAGs and system tests using this API have been updated to explicitly use `XComArg(operator, key=...)` instead.

~TODO:~
~Add docs demonstrating the use of this property with mapped tasks.~ May be handled in #25617

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

josh-fell committed:
Add `output` property to common operator implementation

uranusjr committed:
Move 'output' to operator implementations

This makes things easier for type checkers.

josh-fell committed:
Fix mypy errors in example DAGs and system tests

uranusjr commented on airflow/example_dags/example_xcom.py:
```diff
@@ -79,8 +79,8 @@ def pull_value_from_bash_push(ti=None):
         bash_command='echo "bash pull demo" && '
-        f'echo "The xcom pushed manually is {bash_push.output["manually_pushed_value"]}" && '
-        f'echo "The returned_value xcom is {bash_push.output}" && '
+        f'echo "The xcom pushed manually is {XComArg(bash_push, key="manually_pushed_value")}" && '
+        f'echo "The returned_value xcom is {XComArg(bash_push)}" && '
```
The latter one can still use `output` although using `XComArg` for both is probably more consistent…?

josh-fell replied:
Yeah, it might be better to be consistent. There are some other places I need to clean that up to use `XComArg()` throughout rather than `.output` for now. Is that cool?

josh-fell replied:
Although maybe as long as each file is consistent that's fine. `.output` is still a valid and useful API. I don't think we want to rip it out of all of the examples.

uranusjr commented on airflow/providers/amazon/aws/example_dags/example_ecs.py:
```diff
@@ -99,18 +100,20 @@
     )
     # [END howto_operator_ecs_register_task_definition]
 
+    registered_task_definition = cast(str, register_task.output)
```
It’s probably a good idea to add a Mapy plugin for this… I’ll look into it after this is merged.

josh-fell replied:
Great! I was going to ask if there was a better way than doing all of these `cast()` calls. Had to be something we could do with mypy specifically.

josh-fell committed:
fixup! Fix mypy errors in example DAGs and system tests

josh-fell committed:
Make `XComArg` import compatible with Airflow 2.2

Importing `XComArg` directly from the `airflow` namespace wasn't available until 2.3. Providers only have a minimum req of 2.2. The previous import method would have implicitly required a minimum Airflow 2.3 version.

josh-fell committed:
Resolve conflicts
